design/design-doc-agent.md

```diff
@@ -0,0 +1,586 @@
+    it can be mounted to `Pod`s. Read-only.
+* Push
+  * `*Requirement` custom resources so that infrastructure can be requested.
+    Read and write permissions in a specific namespace in the central cluster will
+    be needed.
+
+### RBAC
+
+As we have two different Kubernetes cluster, there will be two separate security
+domains and because of that, the `ServiceAccount`s in the remote cluster will
+not be able to do any operation in the central cluster. Since the entity that
+will execute the operations in the central cluster is the agent, we need to
+define how we can deliver the necessary credentials to the agent so that it can
+connect to the central cluster. Additionally, it will need some permissions to
+execute operations in the remote cluster like `Secret` and
+`CustomResourceDefinition` creation. We will look at how the agent will be
+granted permissions to do its job in two separate domains with different
+mechanisms.
+
+#### Authenticating to The Central Cluster
+
+In order to execute any operation, a `ServiceAccount` needs to exist in the
+central cluster with appropriate permissions to read during pull and write
+during push operations while synchronizing with central cluster. Since the agent
+is running in the remote cluster, the credentials of this `ServiceAccount` will
+be stored in a `Secret` in the remote cluster. Alongside the credentials, the
+agent needs to know the namespace that it should sync to in the central cluster.
+
+The way to have the agent find the right `Secret` to use in order to
+authenticate to the central cluster for synchronization of a given requirement
+will consist of two steps:
+
+* Look for specific annotations on the requirement type to specify what `Secret`
+  needs to be used:
+  * `agent.crossplane.io/credentials-secret-name: sa-secret` for the full
+    kubeconfig to connect to the central cluster.
+  * `agent.crossplane.io/namespace: my-app-1` for the namespace it should sync
+    requirements and their secrets.
+* If no annotation is given, then use the defaults.
+  * Installation procedure of the agent will allow to specify a default secret
+    and namespace during installation.
+
+By this override behavior, it is possible to supply credentials of another
+namespace or `ServiceAccount` for a specific `*Requirement` resource. So, users
+can target different namespaces in the central cluster from different namespaces
+in the remote cluster, having the ability to have the granularity of RBAC and
+synchronization as however they'd like.
+
+> Though the override behavior also paves the way to target a completely
+> different central cluster, too, which may break the setup since CRDs in that
+> other central cluster could be different. We accept this risk as it would have
+> to be a deliberate choice of the user.
+
+#### Authorization
+
+##### Remote Cluster
+
+Since there will be one agent for the whole cluster, its own mounted
```

Comment:
It feels like agent could connect safely to one cluster without becoming a mess, but could have n scoped connections for different namespace mappings, or could have one privileged connection with n mappings.

Reply:
Sorry, I couldn't get what you mean. Could you elaborate?

Reply:
Sorry for the confusion. Originally when we were talking about supporting annotations that pointed to a secret and remote namespace, I was commenting on how this model might be quite complex if the agent connected arbitrarily many times between any central and remote namespaces. I think it's clarified now already with the namespace annotation updates you've made.